src/tools/db-upgrade/DbUpgrader.cpp

```diff
@@ -0,0 +1,461 @@
+            if (!newIndexKey.empty()) {
+                data.emplace_back(std::move(newIndexKey), "");
+            }
+        }
+    }
+}
+
+std::string DbUpgrader::indexEdgeKey(PartitionID partId,
+                        RowReader* reader,
+                        VertexID& svId,
+                        EdgeRanking rank,
+                        VertexID& dstId,
+                        std::shared_ptr<nebula::meta::cpp2::IndexItem> index) {
+    auto values = IndexKeyUtils::collectIndexValues(reader, index->get_fields());
+    if (!values.ok()) {
+        return "";
+    }
+    return IndexKeyUtils::edgeIndexKey(spaceVidLen_,
+                                       partId,
+                                       index->get_index_id(),
+                                       svId,
+                                       rank,
+                                       dstId,
+                                       std::move(values).value());
+}
+
+void DbUpgrader::doProcessAllTagsAndEdges() {
+    auto ret = buildSchemaAndIndex();
+    if (!ret.ok()) {
+         LOG(INFO) << "Build schema and index failed";
+         return;
+    }
+
+    for (auto& partId : parts_) {
+        auto prefix = NebulaKeyUtilsV1::prefix(partId);
+        std::unique_ptr<kvstore::KVIterator> iter;
+        auto retCode = engine_->prefix(prefix, &iter);
+        if (retCode != kvstore::ResultCode::SUCCEEDED) {
+            LOG(INFO) << "Space id " << spaceId_ << " part " << partId
+                      << " no found!";
+            continue;
+        }
+        std::vector<kvstore::KV> data;
+
+        TagID                                lastTagId = 0;
+        int64_t                              lastVertexId = 0;
+
+        int64_t                              lastSrcVertexId = 0;
+        EdgeType                             lastEdgeType = 0;
+        int64_t                              lastDstVertexId = 0;
+        EdgeRanking                          lastRank = 0;
+
+        while (iter && iter->valid()) {
+            auto key = iter->key();
+            if (NebulaKeyUtilsV1::isVertex(key)) {
+                auto vId = NebulaKeyUtilsV1::getVertexId(key);
+                auto tagId = NebulaKeyUtilsV1::getTagId(key);
+                auto version = NebulaKeyUtilsV1::getVersion(key);
+
+                auto it = tagSchemas_.find(tagId);
+                if (it == tagSchemas_.end()) {
+                    // Invalid data
+                    iter->next();
+                    continue;
+                }
+                if (vId == lastVertexId && tagId == lastTagId) {
+                    // Multi version
+                    iter->next();
+                    continue;
+                }
+
+                // auto strVid = folly::to<std::string>(vId);
+                auto strVid = std::string(reinterpret_cast<const char*>(&vId), sizeof(vId));
+                if (!isValidVidLen(strVid)) {
```

Comment:
I think we don't need to check `isValidVidLen`

Reply:
Yeah, Are all int vid， could remove